libraries/state_history/include/eosio/state_history/log.hpp

```diff
@@ -110,22 +167,41 @@ class state_history_log {
       }
 
       if (block_num < _end_block)
-         truncate(block_num);
-      log.seek_end(0);
+         truncate(block_num); //truncate is expected to always leave file pointer at the end
+      else if (!prune_blocks)
+         log.seek_end(0);
+      else if (prune_blocks && _begin_block != _end_block)
+         log.seek_end(-sizeof(uint32_t));  //overwrite the trailing block count marker on this write
+
+      //if we're operating on a pruned block log and this is the first entry in the log, make note of the feature in the header
+      if(prune_blocks && _begin_block == _end_block)
+         header.magic = ship_magic(get_ship_version(header.magic), ship_feature_pruned_log);
+
       uint64_t pos = log.tellp();
       write_header(header);
       write_payload(log);
       uint64_t end = log.tellp();
-      EOS_ASSERT(end == pos + state_history_log_header_serial_size + header.payload_size, chain::plugin_exception,
-                 "wrote payload with incorrect size to ${name}.log", ("name", name));
-      log.write((char*)&pos, sizeof(pos));
+      fc::raw::pack(log, pos);
 
-      index.seek_end(0);
-      index.write((char*)&pos, sizeof(pos));
+      fc::raw::pack(index, pos);
       if (_begin_block == _end_block)
-         _begin_block = block_num;
+         _index_begin_block = _begin_block = block_num;
       _end_block    = block_num + 1;
       last_block_id = header.block_id;
+
+      if(prune_blocks) {
+         //only bother to try pruning every 4MB written. except for when the config is set very low (mainly for unit test purposes)
+         uint64_t bother_every = 4*1024*1024;
+         if(*prune_blocks < 16)
+            bother_every = 4096;
+
+         const uint64_t mask = ~(bother_every-1);
+         if((pos&mask) != (end&mask))
+            prune();
```

Comment:
> `prune` won't be called if each entry is less than 4 MiB no matter how many such entries are written. There is no accumulated tracking across `write_entry` calls.

Why wouldn't it? Let's consider bother_every=256 which makes mask 0xFF, and writing 4 entries that are all smaller than than 256.
```
pos: 0x000 end:0x040   maskedpos: 0x000 maskedend: 0x000
pos: 0x040 end:0x0a0   maskedpos: 0x000 maskedend: 0x000
pos: 0x0a0 end:0x0e0   maskedpos: 0x000 maskedend: 0x000
pos: 0x0e0 end:0x120   maskedpos: 0x000 maskedend: 0x100  maskedpos != maskedend, so prune()
```

Reply:
Yeah all the pruning options should be bundled in to a config struct. I'll get to that tomrrow.

Reply:
Yeah, I see. But it seems there is still a flaw if it is aligned in an unfortunate way. For example:

```
pos: 0x000 end:0x038   maskedpos: 0x000 maskedend: 0x000
pos: 0x040 end:0x078   maskedpos: 0x000 maskedend: 0x000
pos: 0x080 end:0x0b8   maskedpos: 0x000 maskedend: 0x000
pos: 0x0c0 end:0x0f8   maskedpos: 0x000 maskedend: 0x000
pos: 0x100 end:0x138   maskedpos: 0x100 maskedend: 0x100
```

I think that can be resolved by ensuring the end also includes the 8 bytes suffix so that the tracked data ranges always remain contiguous.